Update to node 18

manotas-scv opened the pull request:
Same as https://github.com/pola88/express-routes-controllers/pull/18 but includes package name change.

- Changed package name to @avi-on/express-routes-controllers
- Package is now an ESM module
- Update node version to hydrogen (18)
- Rest.resource and Rest.resources are now async
- Use es6 features
    - Rest, Routes, Resource and Resources are now Classes
    - Use native versions of lodash methods when available
- Replaced node-jasmine with the latest jasmine version
- Updated eslint and changed rules
- Added prettier and npm run format

manotas-scv committed:
chore: remove obsolete .travis.yml

manotas-scv committed:
chore: replace jasmine-node with jasmine

manotas-scv committed:
chore: refactor to use classes

manotas-scv committed:
chore: refactor slash regexp to utils

facundopereztomasek commented on lib/index.js:
```diff
@@ -83,224 +83,237 @@ If not version is specified or wildcard (['*']), it would be enought with <contr
-    _.each(settings, function(value, key) {self.settings[key] = value; });
+export default class Rest {
+  constructor(settings) {
+    this.settings = {};
+    this.routes = [];
+    this.nestedPath = null;
+    this.subfolder = null;
+
+    settings.versioning = settings.versioning || {};
+    this.vnd_header = settings.versioning.header?.toLowerCase() || 'accept';
+    // Example: /vnd.mycompany.com\+json; version=(\d+)(,|$)/;
+    this.vnd_grab = settings.versioning.grab;
+    this.vnd_error = settings.versioning.error || 406;
+
+    delete settings.versioning;
+    this.settings = { ...settings };
```
```suggestion
    this.settings = {...settings};
    this.routes = [];
    this.nestedPath = null;
    this.subfolder = null;

    this.vnd_header = settings?.versioning?.header?.toLowerCase() || 'accept';
    // Example: /vnd.mycompany.com\+json; version=(\d+)(,|$)/;
    this.vnd_grab = settings?.versioning?.grab;
    this.vnd_error = settings?.versioning?.error || 406;
```

facundopereztomasek replied:
Make sense?

manotas-scv replied:
may be necessary to `delete this.settings.versioning`

facundopereztomasek replied:
why?

manotas-scv replied:
To maintain the old behavior.
The previous version extracted things from the settings.versioning into this, and then deleted that prop from settings.

Afaik the only setting in use is `this.settings.controllers`, so it shouldn't hurt to leave it as you said.

Another option would be to not save the settings object, but to change the behavior to do `this.base_dir = settings.controllers`, which is the only current use for the settings.

manotas-scv replied:
What do you think about:
```js
  constructor(settings = {}) {
    this.routes = [];
    this.nestedPath = null;
    this.subfolder = null;

    this.base_dir = settings.controllers;

    const versioning = settings.versioning || {};
    this.vnd_header = versioning.header?.toLowerCase() || 'accept';
    // Example: /vnd.mycompany.com\+json; version=(\d+)(,|$)/;
    this.vnd_grab = versioning.grab;
    this.vnd_error = versioning.error || 406;
  }
```

facundopereztomasek replied:
Coolllll

facundopereztomasek commented on lib/index.js:
```diff
@@ -83,224 +83,237 @@ If not version is specified or wildcard (['*']), it would be enought with <contr
-    } else if(_.isUndefined(options)) {
+    } else if (_.isUndefined(options)) {
       options = {};
     }
```
```callback || ([options={}, callback] = [callback, options])```

facundopereztomasek replied:
to avoid all the if/else

facundopereztomasek replied:
We may have problems with this case `resource('my resource', {})`
But we can check
```if (typeof callback === 'function') callback()```

manotas-scv replied:
How about:
```js
  async resource(name, options = {}, callback) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
```

This may be longer, but It's more straightforward

facundopereztomasek replied:
great!

facundopereztomasek commented on lib/index.js:
```diff
@@ -83,224 +83,237 @@ If not version is specified or wildcard (['*']), it would be enought with <contr
     resource.init(controller);
 
-    if(typeof callback === 'function') {
+    if (typeof callback === 'function') {
```
at this point callback is a function or undefined, right?

manotas-scv replied:
unless they passed garbage, yes

facundopereztomasek commented on lib/index.js:
```diff
@@ -83,224 +83,237 @@ If not version is specified or wildcard (['*']), it would be enought with <contr
-      } else {
-        throw new Error('Header doesn\'t match RegExp');
-      }
+    if (!(match && version)) {
```
🙌

facundopereztomasek commented on lib/resource.js:
```diff
@@ -1,36 +1,31 @@
+    switch (action) {
+      case 'new':
+      case 'edit':
+        result = addTrailingSlash(result);
```
🥳

manotas-scv committed:
chore: refactor lib/index.js

manotas-scv committed:
chore: refactor lib/routes.js

manotas-scv committed:
feature: ESM + Node 18

manotas-scv committed:
chore: rename package

manotas-scv committed:
fix: handle missing options on ESM

manotas-scv committed:
chore: update eslint + prettier + rules

manotas-scv committed:
chore: Release version 1 & changelog

The change to ESM, and requiring await on resource/resources are breaking changes. And the API seems stable enough to releasea v1
